clippy_lints/src/cognitive_complexity.rs

```diff
@@ -48,190 +52,258 @@ impl LintPass for CognitiveComplexity {
+
+/*
+    Notes on debugging this (14/4):
+    // tests/ui/cognitive_complexity.rs
+    * Check why is 330 not being flagged (it has CoC 0 atm)
+*/
+
+impl<'ast> Visitor<'ast> for CoCHelper {
+    /*
+    # Implemented here:
+
+    ## Nesting Structures
+        IMPORTANT: ExprKind::Block(..)
+        already covers all cases.
+    ## (Nesting-Dependent) Increments
+        if
+        match
+        for, while, loop
+    ## (Nesting-Independent) Increments
+        break, continue
+        Sequences of binary logical operators
+        Recursive function calls
+        Macro calls
+    */
+
+    fn visit_expr(&mut self, ex: &'ast Expr) {
+        match ex.node {
+            // Nesting-Increasing (the one and only)
+            ExprKind::Block(..) => {
+                self.push_nesting();
+                walk_expr(self, ex);
+                self.pop_nesting();
+            },
+
+            ExprKind::Closure(.., _) | ExprKind::IfLet(..) => {
+                // If Let is free of own increment.
+                // This is because it's a language construct
+                // that's specifically designed to save on complexity.
+                walk_expr(self, ex);
+            },
+
+            // FIXME (FAR FUTURE): make a separate, documented case, for recursive calls,
+            // such that it's treated differently from function or method calls.
+            ExprKind::Call(..) | ExprKind::MethodCall(..) => {
+                self.score_ni_structure();
+                walk_expr(self, ex);
+            },
+
+            // Nesting-Dependent
+            ExprKind::If(..)
+            | ExprKind::Match(..)
+            | ExprKind::ForLoop(..)
+            | ExprKind::While(..)
+            | ExprKind::Loop(..) => {
+                // (For the IF-Case)
+                // Important: this pays for one "if" and one "else".
+                // Every "if" in an "else if" comes here again to pay
+                // for itself and its subsequent else.
+                self.score_nd_structure();
+                walk_expr(self, ex);
+            },
+
+            // Nesting-Independent
+            ExprKind::Mac(ref mac) => {
+                self.visit_mac(mac);
             },
-            ExprKind::Call(ref callee, _) => {
-                walk_expr(self, e);
-                let ty = self.cx.tables.node_type(callee.hir_id);
-                match ty.sty {
-                    ty::FnDef(..) | ty::FnPtr(_) => {
-                        let sig = ty.fn_sig(self.cx.tcx);
-                        if sig.skip_binder().output().sty == ty::Never {
-                            self.divergence += 1;
+
+            // Nesting-Independent
+            ExprKind::Break(_, _) | ExprKind::Continue(_) => {
+                self.score_ni_structure();
+            },
+
+            // When boolean operators change, we add 1 to the score.
+            ExprKind::Binary(binop, ref l_ex, ref r_ex) => {
+                // A logical operator
```

Comment:
This essentially ignores the complexity of `(foo && bar) | (a && b)`.

Reply:
Wait. Is that a legal operation? And does it make sense? Aren't you combining logic semantics with bitflag semantics there? Would that snippet be used anywhere? Sounds like it's equivalent to the logical OR of both, is it not?
We should probably implement a separate branch for bitwise operations, but the relationship between them and logic ops is not that clear to me here...🤔 please explain

Reply:
bitops and bool ops are equivalent except for one thing: bool ops only evaluate their rhs if the lhs doesn't make that unnecessary.

So `foo && bar()` does not evaluate `bar` if `foo` is `false`. `foo & bar()` always evaluates `bar`.

All bitops apply to booleans, too.

Reply:
Ohh. Ohhhhhh. This changes things. I see.
Note to my future self: check the associativity of bitwise xor, because it could change the kinds of asts we might be expecting.

Reply:
I've implemented it, please do take a look, because I've (a) refactored it to use one general method instead of two, (b) implemented it for the 5 binary logical operators, and (c) turned the documentation up to eleven